### ISSUE TITLE
CPU hogging when typing in Queries

### ISSUE THREAD

**Issue by neuro-sys** [grafana/grafana#21893]
**What happened**:

On keypress, the grafana dashboard hogs the CPU to a halt for long periods of time.

**What you expected to happen**:

A responsive UI non-blocked by IO or number crunching operations.

**How to reproduce it (as minimally and precisely as possible)**:

At least for me, typing in queries into input box triggers the issue.

![image](https://user-images.githubusercontent.com/648622/73638796-c516a380-4673-11ea-9d8f-feb56c95f8ad.png)

**Anything else we need to know?**:

**Environment**:
- Grafana version: Grafana v6.2.2 (07540df)
- Data source type & version: Not sure, there's a bunch of datasources defined, will report if needed.
- OS Grafana is installed on: 
- User OS & Browser: ArchLinux & Chromium
- Grafana plugins: Polymap, Statusmap
- Others:

**Comment by neuro-sys** [comment]
In case it's useful:

![image](https://user-images.githubusercontent.com/648622/73639547-47539780-4675-11ea-8b39-a5fe4b38fff3.png)

**Comment by peterholmberg** [comment]
Thank you for reporting this @neuro-sys ! Is it possible for you to upgrade to our latest version (6.6) and see if this issue is still there?

Also, you mention that the grafana dashboard freeze when you type in a query input field, however there's no query input field in the dashboard view. Could you help me understand what you mean by this?

**Comment by torkelo** [comment]
What data source are you using when this happens? Are you typing in the query editor?

**Comment by neuro-sys** [comment]
@peterholmberg 
> Thank you for reporting this @neuro-sys ! Is it possible for you to upgrade to our latest version (6.6) and see if this issue is still there?

I'll ask the guys who set it up, might take a little while.

> Also, you mention that the grafana dashboard freeze when you type in a query input field, however there's no query input field in the dashboard view. Could you help me understand what you mean by this?

> Are you typing in the query editor?

It's the panel editing view (I'm not 100% sure on the terminology). It is reached by clicking edit on the tooltip of Dashboard panels.

> What data source are you using when this happens?

It is Prometheus.

**Comment by neuro-sys** [comment]
@peterholmberg 

We upgraded to 6.6, the issue persists.

**Comment by Daegalus** [comment]
To add a bit more detail on @neuro-sys posted (My team handles the Grafana deployments), we were on 6.2.2, and we originally upgraded to 6.5.3, but had massive performance issues related to large data sets (forget the issue number, but it was fixed and merged into 6.6 but was not put in the changelog). So we downgraded to 6.2.2 again, and waited. 6.6 released and we thought it would be fine, but are now hitting this problem, so we had to downgrade again. We have large Prometheus data sets.

Editing queries bogs down the UI considerably and is impacting our users, hence we had to downgrade once more.

The issue is, we want to use plugins that require newer versions than 6.2.2 but we have yet to have a successful upgrade that doesn't make Grafana unusable.

**Comment by torkelo** [comment]
It's strange, we use Prometheus internally with large data sets and no other reports like this. 

Is it when editing Prometheus queries when you see this issue? Can you do page reload with chrome dev tools network tab open, investigate the Prometheus queries, what ar they and what do they return?

**Comment by neuro-sys** [comment]
Today when I fiddle with the same queries (it's back to version 6.2.2 at the moment) I cannot reproduce the issue, and it works just fine without freezes. I am not sure how to proceed at this point as I cannot reproduce it. We could close this issue unless you guys have anything you can do about it.

Just to recap, this issue triggered intermittently but fairly frequently while editing the queries in the query editor (as it triggers an update process after a delay on keypress) resulting in the complete UI freeze for some seconds from a couple to 10s of seconds.

What I did not mention is that after the freeze is gone, it would sometimes insert some piece of text into the query editor at the cursor I was typing. I did not take a screenshot or recognize what they are, but looked like some response data. I would just delete them everytime this happens. I am on Arch Linux with Chromium, so it could be my system is just acting weird.

**Comment by Daegalus** [comment]
I will try to get you some data next week. Right now we are downgraded, and this only happens in production, so we would need to temporarily upgrade to 6.6 again in production to get you the information. Or create a second prod Grafana and add all our sources again, either way, we need to do some prep to get you the data.

**Comment by s-h-a-d-o-w** [comment]
I'm wondering whether this could be related to https://github.com/grafana/grafana/issues/22702.

When this happens again, could you please open the network tab of your browser's devtools, filter by `XHR` and let us know whether there are requests that return something that's larger than 100 KB? (Ideally also a screenshot if it is indeed the case.)
I realize that in the issue I linked to, "hundreds of MBs" is mentioned. Still, when I look at one of our own machines that offers ~3000 metrics, no request exceeds 100 KB. And so it would be helpful in figuring out whether these two issues are indeed related and what the amount where things get problematic is.

**Comment by s-h-a-d-o-w** [comment]
Another thing that I found is that very long label names (in the magnitude of hundreds of thousands or even millions of characters) definitely lead to signifcant delays per key press.
Could you please let us know if that is the case and if so, what order of magnitude is the length of your long label names?

**Comment by s-h-a-d-o-w** [comment]
The requests haven't changed since. It would probably also be a good idea if you typed `{}` to trigger the label name autocomplete and wait for the list of suggestions to pop up, since that also triggers a network request the size of which might be of interest.

**Comment by Daegalus** [comment]
Doing `{}` didnt trigger anything, but typing in  `sum` did. Looks like its transfers 13mb on each request, which expands to 305mb according to the network tab with a total of 23-30 seconds of network time.

[Screenshot](https://ocean.yuli.dev/f/563d59ed7d2a43548ac0/)

**Comment by s-h-a-d-o-w** [comment]
Thank you very much! That's a great first step in us hopefully being able to reproduce the problem. I do have some follow-up questions though:

- When you click on one of those large `values` requests, go to `Preview`, expand `data` and scroll to the bottom => what's the array size?
- With regards to `{}` - that's odd because it should trigger a call to `series`. See my screenshot below for an example that does this - except you would have to do with one of your metrics that has lots of labels (**if** you do have such a case). Because obviously, `prometheus_build_info` doesn't have many. :wink: 

![Screenshot_20200407_144529](https://user-images.githubusercontent.com/16936908/78670733-7fc65a80-78de-11ea-8732-184905bc9551.png)

**Comment by Daegalus** [comment]
So, the data set is so large that neither Firefox or Google would give me the full output. So the convenient CURL command generator in google let me pull the data using that. 

```sh
❯ cat output.json | jq '.data | length'
4042123
```
This is when I start typing almost anything in the query box. I just typed "s" and it pulled the values. and it does it any time i start typing again. Pulling the same 304mb dataset every time. While typing `prometheus_build_info{}` (just to see what happens) it created 11 requests for the full dataset. and locked up my browser until it was done.

**Comment by s-h-a-d-o-w** [comment]
Thanks a lot, I've now been able to reproduce the performance issue from the screenshot posted by @neuro-sys !

As for the other things you described - they were already optimized/handled in some manner along the way (having compared 6.2.2 to current master). For instance, now, it says "Loading metrics..." and if you start typing while it's doing that, no more requests are done. That said, the overall issue of handling huge amounts of metric names still needs to be dealt with.

**Comment by ilietrasca** [comment]
Hi,

As @s-h-a-d-o-w  suggested in #20936 I will add my data below: 

Prometheus data source used:

_Version	2.12.0
BuildDate	20190818-13:53:16
Head chunks	1019858
Head time series	389614_

Grafana version:
![image](https://user-images.githubusercontent.com/24522875/79236714-fe3e7180-7e75-11ea-874b-a1fa10690d9f.png)

After I'm selecting the data source the session simply freeze, screenshot below:
![image](https://user-images.githubusercontent.com/24522875/79236812-1e6e3080-7e76-11ea-8da7-f6a36a70d7f9.png)

After I'm pressing **"Wait"**  the below error is returned and I can start writing the query:
`Dynamic label lookup is disabled for data sources with more than 10000 metrics.`

Screenshot:
![image](https://user-images.githubusercontent.com/24522875/79236946-465d9400-7e76-11ea-9d4c-aca5fbd23678.png)

**Comment by s-h-a-d-o-w** [comment]
Alright everyone, good news - as of 7.0, [there will be an off switch for metrics lookup](https://github.com/grafana/grafana/pull/24034) in the prometheus data source settings!

I will leave this issue open though until it is decided whether we will go ahead with #23811 (for a more sophisticated approach to this problem) or not.

**Comment by ivanahuckova** [comment]
This issue should be removed by:
1. [Add time range parameters to labels API](https://github.com/grafana/grafana/pull/27548) 
2. [Remove running of duplicated metrics query](https://github.com/grafana/grafana/pull/30108)
3. [Improve autocomplete performance and remove disabling of dynamic label lookup](https://github.com/grafana/grafana/pull/30199)

Please reopen if you still experience this issue.
